src/VisualStudio/Core/Def/Implementation/Remote/JsonRpcClient.cs

```diff
@@ -151,15 +138,23 @@ private bool ReportUnlessCanceled(Exception ex, CancellationToken cancellationTo
             return true;
         }
 
-        private void ThrowSoftCrashException(Exception ex, CancellationToken token)
+        private SoftCrashException CreateSoftCrashException(Exception ex, CancellationToken cancellationToken)
         {
+
+            // we are getting unexpected exception from service hub. rather than doing hard crash on unexpected exception,
+            // we decided to do soft crash where we show info bar to users saying "VS got corrupted and users should save
+            // their works and close VS"
+
+            cancellationToken.ThrowIfCancellationRequested();
```

Comment:
this seems weird.  in the codepath to create and throw an exception that tells the user that something went wrong... we potentially cancel from doing that leading to what... just corrupt state that the user doesn't know anything about?  Seems very wonky.

Reply:
All operations that take cancellation token should handle cancellation without corrupting any state regardless of what caused the cancellation. I believe this code is using that fact and pretends remote call failure to be cancellation. Not something I like but that's what it is currently. Not changing that in this PR.